### ISSUE TITLE
Porting steps

### ISSUE THREAD

**Issue by geobra** [geobra/shmong#1]
What works:
- Host client with 1to1 send plain message send and receive
- SFOS startup and connect.

Needs porting:
- xmpp ping in ConnectionManager
- MucHandler
- MamManager
- Omemo
- DiscoInfo
- Presence
- httpUpload (sending / receiving files)

Issues:
- SFOS client crash when entering MessagingPage

@ron282 Any help is appreciated. Just pick something if you want to help and have time.
@anyone else is welcome to also pick a task ;-)

**Comment by ron282** [comment]
@geobra I'm working on it.

**Comment by geobra** [comment]
What are you working on? We must make sure that we are not working on same stuff in parallel.
And we should make a separate PR on each task.

**Comment by ron282** [comment]
@geobra I was stuck with my attempt to create .spec files to compile libomemo-c, qca and qxmpp with sfdk command line but I just solved the issues yesterday. I started to work on the Mam part. I looked at the Omemo encryption but I did nothing for the moment. I saw that we have to provide a storage for encryption data (what was done in axc library). Qxmpp just gives an interface. I agree with your proposal to create distinct PR and I will only work on the ones you assigned to me.

**Comment by geobra** [comment]
Ok, got it. What about if you keep on to focus on Mam and 1to1 Omemo? I will work  on xmpp ping and the crash on SFOS.
I will update the first post in this thread to keep an overview and track the progress.

**Comment by ron282** [comment]
You are talking about crash on SFOS. I solved some. It is working now.

**Comment by geobra** [comment]
Ok, sounds great! I will pick Chatmakers next, if you fixed the crash already.

**Comment by ron282** [comment]
Hi @geobra . I'm still working on porting omemo. I have some issues relate to interoperability with shmoose and may be with third party clients. 
First the xml namespace used in qxmpp is urn:xmpp:omemo and not eu.siacs.conversations.axolotl. This issue is easy to fix:  recompile qxmpp with the right value in QXmppConstants. 
Second is about the way bundles are published. In shmoose/lurch there is one node per bundle with the device id , while qxmpp is putting all bundles under a unique node. Qxmpp seems more compliant with latest standards. 
Therefore what is important is to interoperate.

**Comment by ron282** [comment]
Hi, I have a 2.3.4 version of qca in my repositories that runs with Shmong. When trying to exchange with Conversations, I have a red msg in Conversations: This message is encrypted with OMEMO 2 but could not be decrypted. I don't know if this is an issue with encryption or a lack of support of OMEMO 2 with Conversations.  I need to find another client supporting OMEMO 2 to check if it works fine.  I will  start between 2 Shmong clients.

**Comment by ron282** [comment]
I think I found the main issue. Shmoose and Conversations are relying on OMEMO version 0.3.0 (i.e. publishing devices in eu.siacs.conversations.axolotl.devicelist) while QXmpp is based on  OMEMO 0.8.0 (publishing devices in urn:xmpp:omemo:2:devices, ...).

**Comment by geobra** [comment]
Hm, thats a pitty. Yes, it make sense to implement the latest OMEMO version if one do it from scratch. But to not have compatibility to other clients is not what we want to have. And start to patch qxmpp lib (for older omemo versions) might also have side-effects.

I searched around to try to find out when conversations will update its omemo version. There is one ticket:
https://github.com/iNPUTmice/Conversations/issues/4191
But no information about the way forward with conversations.

So, where to go from here...?!

**Comment by ron282** [comment]
I'm trying to check if I can have Conversations working with Shmong with few patches. We can also asked people involved in QXmpp what is their point of view. Even if OMEMO is still a draft, having a mechanism to ensure interoperability is essential for me to make it successful. On the web site listing OMEMO implementations I didn't see any mention of the OMEMO version but this is an important parameter.

**Comment by geobra** [comment]
Yes, sounds like a good approach. And sooner or later, also Conversations will switch to the new omemo version. Until then, we have a working client with Shmoose. Later on, we could switch to Shmong. And in the mean time, we have enough time to develop it and do the porting. 

Do you know another client which already uses the new omemo version? The one client mentioned in the #4191 ticket is only a windows client.

**Comment by ron282** [comment]
I can see that this OMEMO2 evolution is not going to happen soon in deployed clients: https://logs.xmpp.org/xsf/2020-01-08.

**Comment by geobra** [comment]
This log is quite old from 2020. But, reading this, it sounds more difficult then I thought it would be to go from omemo1 to omemo2 :-(.

**Comment by geobra** [comment]
So, the question is, is it is worth the effort of going the qxmpp way, if it really takes years for the majority of the other xmpp clients to switch to omemo2.

**Comment by geobra** [comment]
This project [1] has implemented both, omemo 1 and omemo 2, in one library. And the library handles both versions transparent to the using app project. But it is python :-(.

[1] https://github.com/Syndace/python-omemo

**Comment by ron282** [comment]
I'm trying to get a  qxmpp working with old omemo implementation.

**Comment by geobra** [comment]
Maybe it make sense to ask the qxmpp project if they already have plans what way to go to get both, omemo1 and omemo2 working within the library. I will do this the next days...

**Comment by ron282** [comment]
Yes. You can ask. I'm still working it. Solving issue after issue.

**Comment by geobra** [comment]
Here is the discussion at the qxmpp project. They have no resources left to work on both omemo versions.
https://github.com/qxmpp-project/qxmpp/issues/515

**Comment by ron282** [comment]
I did all the changes but I'm blocked at libomemo level.  I have an error message: WARNING OMEMO envelope data could not be deserialized. It looks like in libomemo the function pre_key_signal_message_deserialize_omemo fails. May a question of version between version 3 and version 4... I will try to solve it but I'm not an expert.

**Comment by geobra** [comment]
For sure, this is not an easy task. If you want me to also have a look at it, just make a pull request. I can try to give some hints or suggestions.

**Comment by ron282** [comment]
I have a piece of code that works ! There are  tons of edits. I put it in github.  Now I understand a little bit more why Omemo has evolved. Implementation supported by Conversations and others are leaking data.  What is needed now is to check how to support both.

**Comment by ron282** [comment]
For the moment, I will stop to a version that only supports Omemo 3 (legacy version) based on QXMPP and now may be come back to Shmong and the reason that pushes us to look at Qxmpp. Don't hesitate to give your advices about how to write QXmpp changes in  the most maintainable way

**Comment by geobra** [comment]
Nice to read this. Do you have a link to your changes? The last commits in your github repo are from 13th of October...

**Comment by ron282** [comment]
I put put here: https://github.com/ron282/qxmpp/tree/qxmpp-sfos

**Comment by ron282** [comment]
I also put a test version of shmong that works with it here: https://github.com/ron282/shmong. I was able to exchange encrypted messages with Conversations.

**Comment by geobra** [comment]
Sorry for not being responsive the last weeks. There is a lot of workload currently on my desk :-(.

I had a look at your qxmpp changes. Great to hear that it is working as expect. As I don't know which coding style and what quality gates should be applied for the qxmpp project, it probably make sense to create a WIP (work in progress) Pull Request directly to the qxmpp project and ask for a review there.

I will have a look at your shmong branch and test it at the weekend :-).

**Comment by ron282** [comment]
Hi, I found errors related to node deletion in the qxmpp code I added. I also have a better understanding of qxmpp. I hope I will be able to publish soon improvements.

**Comment by ron282** [comment]
@geobra I have a quite stable release. How do you want to proceed to move forward?

    xmpp ping in ConnectionManager [geobra]
    MucHandler [ron282 in progress]
    MamManager [ron282 done need testing]
    Message Carbons [ron282 done need testing]
    Omemo [ron282 done need testing]
    DiscoInfo [ron282 done]
    Presence [ron282 done need testing]
    httpUpload (sending / receiving files) [ron282 done need testing ]
    ChatMarker [geobra]

**Comment by geobra** [comment]
Hi

Great progress and success! You have implemented and achieved a lot of functionality, despite the uncertain start during switch to qxmpp.

To be honest, for me I moved forward to different other things then implementing stuff in C++. I spent not much time the last month and years on actual programming. I would encourage you to take over the new project on your own github space. You are into the actual implementation and you will be the best owner of that client and project.

I can help you (if needed) with knowledge from the past where I started and released versions of Shmoose to get you up and running with a release of Shmong (or how you would like to name it). I can also support you with small implementations. But it does not make sense if I take the 'main developer role' of that new client.

What do you think about this?

**Comment by ron282** [comment]
Hi

Thanks for your reply. I will follow this road. I may ask you few questions and take benefit of your experience and your advises. This is also a hobby for me and I may spend less time in the future. I wasn't expecting  porting to qxmpp would be so long.

**Comment by ron282** [comment]
@geobra 
Thanks for the background. This is interesting for me to know why you developed Shmoose.
I had a look at the thread about Qt 5.15 with Sailfish. It's interesting. I don't know what it means in term of changes on the QML part.
I put on Openrepos a first alpha of Shmong. I'd like to add some tests but I don't know what is the best way to do it since QTest is not supported with Qt 5.6. For the future, I don't know which road to follow: either adding features, either integrating an existing UI (you mentioned this possibly in a past msg). Anyway having your suggestions is good food for thoughts.

**Comment by jlaba** [comment]
Great to here that! Can you add the qca packages to openrepos as well? I really like to test shmong :-)

Thanks for your effort!

onsdag den 17 maj 2023 skrev ron282:

**Comment by ron282** [comment]
@jlaba QCA is already available on my account on Openrepos. May be you are requesting the devel version? I can add it.

**Comment by jlaba** [comment]
I have seen this, but there are no packeges listed below the qca repo. Therefore it couls not be installed. In Storeman the failure message is something like "No package available", when I try to install the package.

Regards!

fredag den 19 maj 2023 skrev ron282:

**Comment by geobra** [comment]
@ron282 
Great news that you managed it to get a first version to Openrepos :-)!
I also tried to install it on my device. And get the same error as @jlaba. The other dependencies (qxmpp, libomeo) are available on your account and filled with packages. The package for qca is missing.

About the Qt 5.15, qml and Kaidan way:
In theory, it should be possible to have Kaidan mainline build for SFOS and the Qt 5.15 package. All the dependencies are now available. For the qml part, it could be possible that the Kaidan Kirigami qml UI just works on SFOS. In practice, there will probably be some pitfalls. Maybe some stuff did not compile with the SFOS toolchain and needs minor patches. Similar to the patch we created for qxmpp. And for sure, the integration into the SFOS notification system must be ported from Shmong. As well as the iphb (to prevent deep sleep of the device) implementation. Having said that, my idea was not to integrate the Kaidan UI in Shmong, but it could also be a path forward to port Kaidan 'as is' to SFOS.

If you want to stick to Shmong, I would recommend to focus on the testing stuff. Why is QTest not available in 5.6? I used it all the time in the past. Or did I miss something? I was a failure not to implement an automated test for every class in the past. You never know if something breaks in the code on new features. Having tests will speedup development and increase quality in the long run. No need to failure prone manual testing any more.

**Comment by geobra** [comment]
@ron282 
The guys from the qt5.15 port now indeed started to build the dependencies for Kaidan to get it build for SFOS.
https://forum.sailfishos.org/t/xmpp-in-sailfish-how-to-start/12960
The most interesting stuff is at the bottom of that thread.

**Comment by ron282** [comment]
@jlaba @geobra I put qca on OpenRepos in my repository. I looked at the bottom of the thread and yes this is great. Now it should be possible to use the Qxmpp version "Omemo 0.3 compatible" with Kaidan.

**Comment by ron282** [comment]
I'm currently trying to compile kaidan with qt 5.6. I have the first half of source files compiling. It requires some changes (QStringView, invokeMethod with function, QRandomGenerator...). Once this done and if i don't hit a wall, i'd like to modify the qml files to use sailfish qml components. And then i will check if it works without taking too much debug time. If not, i will continue on Shmong, trying to keep close to kaidan architecture. I don't have a lot of time, so it will progress slowly. My goal is to get a qxmpp client for sailfish that i can use everyday.

**Comment by poetaster** [comment]
I've testing the shmong client on chum testing and it seems to work. I need some handles to communicate with. All my other contacts are not longer on jabber ;-(
Pushing to chum now.

**Comment by ron282** [comment]
Just curiosity, are your contacts now on what's app, matrix or something else? I have source files from Kaidan from A to M that are compiling. I need N to Z ones to complete.

**Comment by poetaster** [comment]
> Just curiosity, are your contacts now on what's app, matrix or something else?

I hang out on the #sailfishos (poetaster) channel on OFTC irc server and I hang out at poetaster@sailmates.net matrix server ... those two are almost always on. I do also use telegram and signal, but telegram mainly for friends.

> I have source files from Kaidan from A to M that are compiling. I need N to Z ones to complete.

Damn :) I sorta stopped after I noticed all the c++20 stuff creeping in. Although, I guess it's not 'that' much?

**Comment by ron282** [comment]
I have the c++ compiling. I’m adapting the qml to work with Sailfish. I have a couple of pages that are working. But not yet possible to set up a connection.

**Comment by poetaster** [comment]
> I have the c++ compiling. I’m adapting the qml to work with Sailfish. I have a couple of pages that are working. But not yet possible to set up a connection.

If you're going to use the opt 5.15 libs from rinigus and piggz you can't use the sailfish silica stuff. But, I'm not sure if you were intending on shipping ALL the kde libs?

**Comment by ron282** [comment]
You are correct I removed the kde stuff which is mainly used for the qml part. Basically my goal is to keep the c++ part  and to get a Sailfish qml.

**Comment by mazhe** [comment]
oh, and here I was, trying to debug and do some stuff based on your shmong codebase... but a Kaidan based client would make sense to reuse a lot of work, and they seems to have put a lot of work those last years... one fear is that I understood they were not planning any MUC support in kaidan...

**Comment by ron282** [comment]
I can't say when Kaidan for Sailfish will work. Currently it is connecting to xmpp server. I'm debugging... I need to understand how Kaidan is working to debug. When it works I guess  it won't be difficult to add MUC support.

**Comment by ron282** [comment]
Still debugging. The C++ part seems to work fine. Lot of work on the qml side. I have few pages working but nothing to share right now.

**Comment by ron282** [comment]
Hi,

I have a version based on Kaidan that is working for text messages. There is still a lot to do regarding support of Omemo, file exchange, qr code usage, cover page. I'm spending a lot of time on the QML part which needs to be adapted to match Sailfish UI but also to remain close to current code to ease porting new features. I may release something in the coming weeks disabling  non working features.

**Comment by mazhe** [comment]
Congrats and thanks! I'd be happy to see a basic text version (with carbon and mam?), I'm sure features can be reintroduced one at a time -- if I have the time, I'll try to help.

**Comment by ron282** [comment]
@mazhe I put on Openrepos a very first alpha.

**Comment by jlaba** [comment]
Great to here! Tried to install it, but libzxing is missing.

**Comment by ron282** [comment]
@jlaba Oups I'm adding it

**Comment by ron282** [comment]
@jlaba Tell me if it is fixed

**Comment by mazhe** [comment]
> @mazhe I put on Openrepos a very first alpha.

Awesome! Is it buildable through sfdk with your github repo? My XA2 is only armv7...

**Comment by jlaba** [comment]
Yes, it is fixed. Thanks!

onsdag den 15 november 2023 skrev ron282:

**Comment by ron282** [comment]
> > @mazhe I put on Openrepos a very first alpha.
> 
> Awesome! Is it buildable through sfdk with your github repo? My XA2 is only armv7...

i started to work on it. I have a dependency issue to fix. I also own an Xperia 10 to test. Stay tuned.

**Comment by Neustradamus** [comment]
To follow this important ticket :)

**Comment by Neustradamus** [comment]
It is possible to have releases, the developement has been stopped?

It is possible to move this repository in a GitHub organization and update it for the future?

Thanks in advance.

**Comment by mazhe** [comment]
I don't think it has stopped as much as "proof of concept is done, but a path must be found". I've tried to rebase the fork on newer kaidan codebase but it's a lot of work. I think the only way forward would be to at least be able to get the C++ bits merged in Kaidan (and attribution given to @ron282), but I haven't found the time to try to approach them and feel the temperature. Then there's the QML part...

**Comment by ron282** [comment]
I'm working on the common code base between Shmong and Kaidan (QXmpp). This is working but it has some bugs I have difficulties to fix related to Omemo. I put on pause work on Shmong, focusing on porting the full Kaidan to SailfishOS. I have a very first alpha. The C++ code is close  to Kaidan but the QML part needs a complete rework and this is really a hard work for me.  And finally when I look at Kaidan  features compared to Shmong, I'm a little bit disappointed. I will  provide you an armv7 version.

**Comment by geobra** [comment]
See also my comment here:
https://github.com/geobra/harbour-shmoose/issues/140#issuecomment-1872610731

Additional to that:
Kaidan always uses bleeding edge versions of technology. New C++ standards, latests versions of components. Sailfish has a quite old SDK and also quite old software. And the latest problems with the new Jolla company structure is also not a good sign to get this situation better in the near future.

There will always be the need of 'extra work' to be done to backport the Kaidan source to this old ecosystem. Either this must be accepted or other alternative Linux mobile OSes should be used. I had a look at postmarket OS recently. There are Dino and Kaidan already available. But this OS has other glitches like e.g. not a consistent UI. Windows are too broad to fit to the screen.

As of today, there is no perfect solution which fits to all the needs ;-).
